### ISSUE TITLE
Custom chunks losing data in rare cases

### ISSUE THREAD

**Issue by Bobbyharron** [PluginBugs/Issues-ItemsAdder#2403]
### Terms

- [X] I'm using the very latest version of ItemsAdder and its dependencies.
- [X] I am sure this is a bug and it is not caused by a misconfiguration or by another plugin.
- [X] I already searched on this [Github page](https://github.com/PluginBugs/Issues-ItemsAdder/issues) to check if the same issue was already reported.
- [X] I already searched on the [plugin wiki](https://itemsadder.devs.beer/) to know if a solution is already known.
- [X] I already searched on the [forums](https://forum.devs.beer/) to check if anyone already has a solution for this.

### Discord tag (optional)

bobby_harron#7548

### What happened?

i come from this post:
https://github.com/PluginBugs/Issues-ItemsAdder/issues/2398

Since you closed the last post and the issue is not solved and i dont know if you look at closed post i made this one.

I tested it with a new server and only Essentials, Chestcommands fork and Itemsadder (with Lonelibs and ProtocolLib) All up to date, even itemsadder with the new 3.4.0c-beta.

https://www.youtube.com/watch?v=B4Z5uEB5MiA&ab_channel=BobbyHarron

### Steps to reproduce the issue

With all the information above, just make a custom block using real noteblock and right click it.

### Server version

Current: git-Purpur-1927 (MC: 1.19.3)*
Previous: git-Purpur-1858 (MC: 1.19.2)
* You are running the latest version

### ItemsAdder Version

3.4.0c-beta

### ProtocolLib Version

5.0.0 SNAPSHOT b612

### LoneLibs Version

1.0.23

### FULL server log

[latest.log](https://github.com/PluginBugs/Issues-ItemsAdder/files/10944328/latest.log)

**Comment by LoneDev6** [comment]
It might be purpur then.
Try to use Paper and let me know.
https://papermc.io/downloads

**Comment by spiker5032** [comment]
Im also experiencing this issue I am using purpur-1.19.2-1825.jar

**Comment by LoneDev6** [comment]
I didn't find any way to reproduce this issue, so I cannot fix it.
On normal setup the blocks work fine, there is something interfering

**Comment by aaa8881** [comment]
We're expereincing same issue , it happens with customcrop plugin
It usually happens after reboot , and status of block get lost. 
doesn't happen to everyone, so I personally couldn't reproduce by myserlf, but happens quite occassionally.

**Comment by aaa8881** [comment]
I wouldn't say I have never force killed server so far, but issue doesn't happen from process being killed forcibly. 

I've just checked disabling log, but couldn't particularly find error log from IA. 

It definitely doensn't happen to everyone, but definitely happens sometimes. 
These are patterns I found 

1. There are few custom blocks made of IA in my server , including some furnitures, but all reported issues are from customcrop - farmland stuff. not sure if this is coincident. 
2. one it happens, all custom block around it happened get affected. 
3. in many cases, right clicking on the custom block and breaking actually returns correct custom block. However, it still doesn't re-show the texture , but still shown as noteblock 

hope it helps to get some more idea

**Comment by LoneDev6** [comment]
> 3. in many cases, right clicking on the custom block and breaking actually returns correct custom block. However, it still doesn't re-show the texture , but still shown as noteblock

in "many" or in ALL cases? This is important.

**Comment by aaa8881** [comment]
> > 3. in many cases, right clicking on the custom block and breaking actually returns correct custom block. However, it still doesn't re-show the texture , but still shown as noteblock
> 
> in "many" or in ALL cases? This is important.

ALL that I happened to check in game. 

But I couldn't check in game everything because some poeple just removed noteblocks first and then report issue.

**Comment by LoneDev6** [comment]
Note: seems to be related https://forum.devs.beer/d/311-note-blocks-ploblem

**Comment by LoneDev6** [comment]
I tested a lot with the latest ItemsAdder version 3.4.1 and I cannot reproduce the issue.
If you shutdown the server normally ItemsAdder will correctly save blocks data.

**Comment by LoneDev6** [comment]
Are your bugged regions noted in the log files to be converted? Check your old logs.
Something like that:

![image](https://user-images.githubusercontent.com/27242001/228177360-eb7f0d02-4467-45a0-b8dc-30b6926004bb.png)

Let me know.

**Comment by aaa8881** [comment]
I'm actually still in 3.3.1 so, may need to upgrade to 3.4.1 and see if issue persists. will update today. will let u know if issue persists.

**Comment by aaa8881** [comment]
> News?

I updated to 3.4.1 , fortunately most custom blocks nbt seem to be updated correctly, but above issue still persists.

mostly happen from customcrops, but I think it's purely because there are just too many custom crops in the server.

1. custom block made of noteblock sometimes doesn't yield anything on break
2. custom block made of itemframe sometimes doesn't yield anything on break

However, it seems like incidents that custom block made of noteblock becomes pure noteblock are either greatly reduced or solved. I sometimes encountered claim their blocks are having trouble, but it's unclear if that happened before 3.4.1 update or just lying so. at least incident number has dropped a lot.

**Comment by LoneDev6** [comment]
This issue report is very strange since the amount of blocks cannot have anything to do with blocks data lost.
Also I don't see how crops can be the cause.

Maybe the custom crops plugin you are using is doing some strange stuff while the plugin is saving blocks? This could be the cause of the break, maybe.

**Comment by Xiao-MoMi** [comment]
Maybe replacing the block (CustomBlock.remove(), CustomBlock.place()) in unloaded chunks with high server load(aaa8881 is running a server with about 200-300 players) could cause that. This is the only thing I can think of because I have never met this problem.

**Comment by aaa8881** [comment]
problem is.... it is reproduced in my server with very high chance... encountering at least 10 tickets of people claiming pot (customcrop custom block) is changing to noteblock. what can I check to help you?

**Comment by Bobbyharron** [comment]
i know its late for this answer but, i managed to fix it by changing to false the disable-REAL option, inside the config.yml. Its stupid cause the blocks i use are REAL-NOTE, not REAL, but it just worked

**Comment by LoneDev6** [comment]
@Xiao-MoMi said https://github.com/PluginBugs/Issues-ItemsAdder/issues/2602#issuecomment-1537045045 :
> According to the previous issue similiar to this, the common that can be found is that there are many players on the server, which also means that there are more custom blocks, and more crop blocks as well. I personally think that the reason why it was not possible to reproduce before was because the concurrency of the operation was not high enough.

The plugin should handle concurrency just fine, I can try to investigate again but I don't understand how concurrency can be the cause. I mean concurrency issues should cause 1-2 blocks lost, not an entire region.

**Comment by aaa8881** [comment]
It is frustrating issue. what can I help to give information so that you can figure out the issue better?

**Comment by LoneDev6** [comment]
Yes I totally understand and I'd love to fix it as soon as possible.

I am thinking about some questions..
1. is the issue occurring in [REGIONS](https://minecraft.fandom.com/wiki/Region_file_format) (32x32 chunks) where there are a lot of players?
2. is it happening on regions where there is only a little group of players playing?
3. is it happening always in the same regions?
4. do you get any error?
5. are blocks corrupting after some seconds from the placement or were the corrupting blocks placed long time ago?
6. do corruption stop if you remove the crops plugin from your server temporarily?

**Comment by aaa8881** [comment]
I can do most of them, but need clarifications on few

5. how long is ``long time ago`` ?
6. this is probably not easy thing to do at this moment.

**Comment by LoneDev6** [comment]
1 are you 101% sure that the corruption happens on a SINGLE chunk at a time and not in the whole REGION?
4 sometimes it happens also to blocks placed 1-2 minutes ago? 30 seconds ago? or only blocks placed days ago or hours ago?

**Comment by LoneDev6** [comment]
> 6. do corruption stop if you remove the crops plugin from your server temporarily?

This is a crucial test.
I need to understand if the cause is the other plugin using the ItemsAdder blocks API in wrong moments causing access to the chunk data when it shouldn't.

**Comment by aaa8881** [comment]
1. no , I may need to double check them . we're still collecting & restoring people whose custom block is corrupted, so I'll make sure if it happens in chunk or region. Let me collect every single information for past few days and from now, and let you know

here are some interesting symptoms I've discovered and checked
- I said "every single block in chunk" , but it's not true. actually some are corrupted some are not in same chunk, or same region
- some blocks actually "return" to correct status when something is placed adjacently. 

4. let me check that again
6. would placing custom block in server that does not have customcrop plugin , and see if data ever get corrupted satisfy this test? Currently, only "survival" type server have customcrops, and "regen" type servers do not have customcrops. however, ia is same, so I can still place ia blocks in "regen" servers as well. 

Lastly, I couldn't really find any way to contact you privately, so let me speak this for just one time. I'm willing to support you economically if I can get some prioritized support on this issue. this issue is generating at least 100 tickets (no exaggeration) per day, and bottlenecking my community development (and life). I already added you in discord. signition#0001 , I know it could be insulting depends on person, but please consider this not to be rude, but as desperation. will delete this comment after you read.

**Comment by LoneDev6** [comment]
> would placing custom block in server that does not have customcrop plugin , and see if data ever get corrupted satisfy this test?

Yes as soon as the other server has lot of players playing and placing blocks as it seems to be the cause (as you and Xiao suspect).

> I already added you in discord. 
Sorry this is my current situation on Discord, I don't check the friend requests.
![image](https://github.com/PluginBugs/Issues-ItemsAdder/assets/27242001/7c9edfeb-adc7-46ad-a553-9aef18db04b3)

I now will accept your request so that we can get into more detail about your particular server setup and see if it's actually an itemsadder bug or an incompatibility.

**Comment by aaa8881** [comment]
ok not sure what's wrong, but there are many things wrong
```
  blocks:
    log-mined-unknown-blocks: true
    log-convert-vanilla-blocks: true
    log-save-load:
      enabled: false
      min-ms: 20
    log-fix-bugged-blocks: true
    log-unusual-block-break: true
```

these options are enabled
there are 750k lines of ItemsAdder debug message for past 8 hours ( 8AM to 4PM) in one bukkit instance. 

1. ``[ItemsAdder] [mFixed glitched block at 1309.0, 82.0, 5279.0[m``  
A LOT OF THESE in log 
I looked into those positions, they're mostly fine - probably ia fixed it, or not sure.
it happens in bulk - seems like it's triggered by chunk loading or unloading , affects custom blocks in a chunk.

2. ``somehow tried to break a custom block even when not normally possible. They are probably hacking or using Minecraft Bedrock client (Geyser).[m``
This seems to be happening when player is doing something with customcrop blocks, not always but at some incidents.

**Comment by aaa8881** [comment]
I have placed multiple custom blocks in a region where lots of player comes by - at least around 10 players are always around, sometimes up to 50 (concurrent) , so a lot of chunk loading / unloading come and go - it's fine without customcrops. no blocks are corrupted

**Comment by Xiao-MoMi** [comment]
I think the reason for the corruption is that two different threads simultaneously used the IA API to place/remove blocks at the same location. Even if there are many players, their behavior is still on the main thread so it should not cause the corruption. I previously released a patch that refined the task precision of the thread pool used by CustomCrops from **tick** to **ms**. According to a user's feedback, the corruption has reduced.

**Comment by LoneDev6** [comment]
So the issue seems to rely on CustomCrops plugin accessing the ItemsAdder API?
ItemsAdder blocks API should be thread safe, I coded it with concurrency in mind so I don't think CustomCrops accessing the API on different threads is an issue.

**Comment by LoneDev6** [comment]
> 1. ``[ItemsAdder] [mFixed glitched block at 1309.0, 82.0, 5279.0[m``  
A LOT OF THESE in log 
I looked into those positions, they're mostly fine - probably ia fixed it, or not sure.
it happens in bulk - seems like it's triggered by chunk loading or unloading , affects custom blocks in a chunk.

To clarify what ItemsAdder considers a "glitched block": it is a block which was placed by Minecraft vanilla behaviour or by another world generator plugin and it has some custom data properties. The block will probably mistakenly have the texture of a custom block, so ItemsAdder fixes that by setting it to the default vanilla blockstate values.
For example mushroom blocks spawning with some blockstates would look like custom blocks, which is wrong. ItemsAdder will set the default vanilla blockstate for that mushroom block, this will avoid showing fake custom blocks.
Example: https://itemsadder.devs.beer/faq/blocks-minerals-issues/custom-blocks-glitch-texture/avoid-glitched-blocks
This has nothing to do with the bug reported in this issue report.


> 2. ``somehow tried to break a custom block even when not normally possible. They are probably hacking or using Minecraft Bedrock client (Geyser).[m``
This seems to be happening when player is doing something with customcrop blocks, not always but at some incidents. 

This special case is triggered when a player somehow manages to trigger the block break event for custom blocks.
This event doesn't trigger normally because the player needs to send the block breaking animation states, then the plugin detects the last state and marks the block as broken without calling the block break event.
This eventuality usually happens when a player is cheating or using bedrock edition client (Geyser server).
Seems for some reason CustomCrops plugin is calling the `BlockBreakEvent` for these custom blocks even if they don't need to.
@Xiao-MoMi I'm here to help you fixing it :D 


@Xiao-MoMi 
> According to a user's feedback, the corruption has reduced.
So seems async tasks are the issue?
I think in some occasions CustomCrops tries to set a block while a chunk gets saved by ItemsAdder, so partial data gets lost in between? Maybe? It's strange as the chunk saving method locks the resource until it's all saved, so also the API calls would wait until save is completed before trying to load the chunk again to set a block.

**Comment by aaa8881** [comment]
hmm. let me know if I can help you guys more. waiting 24/7 to provide more stuffs

**Comment by Xiao-MoMi** [comment]
I tried many times, but unfortunately, I couldn't reproduce this problem, even if I was doing stress tests on thousands of crops. It is undeniable that this problem does exist, but the problem has not been located yet theoretically.

I agree with Lonedev's views, which is that it is possible that while unloading blocks and storing data in IA, CustomCrops called the API to place/break block's custom blocks.

**Comment by Xiao-MoMi** [comment]
Some additional information I can provide:

In my code, I listened to the CustomBlockPlace/BreakEvent to record the position of the greenhouse glasses, but would not do further operations on them. So corruption affects the chunk, not a single block.

**Comment by aaa8881** [comment]
Just a couple more information 

Also, this issue especially became more problematic after we change customcrop mode from itemframes to tripwire. Unsure if this could be related , but 

Will be clear, but in item frame mode, custom crops are item frame ; thus entity.
in tripwire mode, custom crops are tripwire; thus not entity.

**Comment by LoneDev6** [comment]
I need to know if the issue is 100% caused by CustomCrops plugin.
Thanks.

**Comment by LoneDev6** [comment]
> how would I ever know that?

I can tell the same thing.

The only way to know that is to disable custom crops plugin temporarily and check if the issue occurs for placed blocks (placed blocks only after the CustomCrops plugin removal).
I have no other way to identify the cause of the issue.
I already asked you about that test: https://github.com/PluginBugs/Issues-ItemsAdder/issues/2403#issuecomment-1540459481

**Comment by aaa8881** [comment]
ok. disabling in one server and leave in other server.

**Comment by aaa8881** [comment]
> Do some tests during 2-3 days. Let me know. Thanks

well, I don't think I would need 2-3 days. there are already lots of reports claiming things are happening. After investigation for a while, it's clear that they are definitely corrupted AFTER customcrops plugin is removed from the sever. I did make sure it's not from they logged in like yesterday, and returned today and claim their data is corrupted. there are many active players who played today, before and after the reboot , and found their custom blocks are corrupted.

**Comment by LoneDev6** [comment]
> and found their custom blocks are corrupted.

Are you 100% sure that the corrupted custom blocks were placed after the removal of CustomCrops?

**Comment by LoneDev6** [comment]
Can I get your full plugins list?

**Comment by LoneDev6** [comment]
> The only method related to custom block is CustomBlock.remove(), CustomBlock.place(). Maybe replacing the block (CustomBlock.remove(), CustomBlock.place()) in unloaded chunks with high server load(aaa8881 is running a server with about 200-300 players) could cause that. This is the only thing I can think of because I have never met this problem.

Checking your code seems you are requesting chunks so chunks should be available without any issue, correct?
https://github.com/Xiao-MoMi/Custom-Crops/blob/93f5acf607c5d391b70fd7ee4419db1777d60615/src/main/java/net/momirealms/customcrops/api/object/world/CCWorld.java#L547
So "Maybe replacing the block (CustomBlock.remove(), CustomBlock.place()) in unloaded chunks with high server load" is not a valid statement I think.

**Comment by LoneDev6** [comment]
Might issues be caused by this part of code?
https://github.com/Xiao-MoMi/Custom-Crops/blob/93f5acf607c5d391b70fd7ee4419db1777d60615/src/main/java/net/momirealms/customcrops/api/object/world/CCWorld.java#L584

Probably setting `enable-corruption-fixer: false` can possibly fix the issue.

**Comment by Xiao-MoMi** [comment]
> Might issues be caused by this part of code? https://github.com/Xiao-MoMi/Custom-Crops/blob/93f5acf607c5d391b70fd7ee4419db1777d60615/src/main/java/net/momirealms/customcrops/api/object/world/CCWorld.java#L584
> 
> Probably setting `enable-corruption-fixer: false` can possibly fix the issue.

If a pot is corrupted, then its id would be "NoteBlock" instead of "namespace:id"
So this might fix some pots
placeNoteBlock() method is actually using CustomBlock.place()

**Comment by Xiao-MoMi** [comment]
> 

chunk.getChunkAtAsync is from paper API.
It should be well handled, and the chunk is loaded when replacing blocks.

**Comment by LoneDev6** [comment]
Might have been fixed by with #3245
Please let me know if it will ever happen again.
